### ISSUE TITLE
Scoring not existing model gives misleading error message

### ISSUE THREAD

**Issue by mplpl** [IBM/nzpyida#70]
When we use a model that does not exist, instead of a clear error message we got an exception on deleting a table. This is a side effect of wrong handling that situation (exception in exception handling).

The sample code:

```
model = DecisionTreeClassifier(idadb, model_name='abcd')
model.score(in_df=train_df, target_column='IS_FRAUD')
```

Error:

```
---------------------------------------------------------------------------
ProgrammingError                          Traceback (most recent call last)
File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:149](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:149), in PredictiveModeling._score(self, in_df, predict_params, target_column)
    147 try:
--> 149     pred_df = self._predict(in_df=in_df, params=predict_params, out_table=out_table)
    151     pred_view, pred_view_needs_delete = materialize_df(pred_df)

File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:111](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:111), in PredictiveModeling._predict(self, in_df, params, out_table)
    110 params['model'] = self.model_name
--> 111 return call_proc_df_in_out(proc=self.predict_proc, in_df=in_df, params=params,
    112     out_table=out_table)[0]

File [~/Documents/work/nzpyida/nzpyida/analytics/utils.py:143](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/utils.py:143), in call_proc_df_in_out(proc, in_df, params, out_table, copy_indexer)
    142 try:
--> 143     out_query = in_df.ida_query(f'call NZA..{proc}(\'{params_s}\')')
    144 finally:

File [~/Documents/work/nzpyida/nzpyida/frame.py:1002](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/frame.py:1002), in IdaDataFrame.ida_query(self, query, silent, first_row_only, autocommit)
    996 """
    997 Convenience function delegated from IdaDataBase.
    998 
    999 Prepare, execute and format the result of a query in a data frame or in
   1000 a tuple. See the IdaDataBase.ida_query documentation.
   1001 """
-> 1002 return self._idadb.ida_query(query, silent, first_row_only, autocommit)

File [~/Documents/work/nzpyida/nzpyida/base.py:1033](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/base.py:1033), in IdaDataBase.ida_query(self, query, silent, first_row_only, autocommit)
   1032 self._check_connection()
-> 1033 return sql.ida_query(self, query, silent, first_row_only, autocommit)

File [~/Documents/work/nzpyida/nzpyida/sql.py:85](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/sql.py:85), in ida_query(idadb, query, silent, first_row_only, autocommit)
     84 elif idadb._con_type == 'nzpy':
---> 85     return _ida_query_NZPY(idadb, query, silent, first_row_only, autocommit)
     86 else:

File [~/Documents/work/nzpyida/nzpyida/sql.py:193](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/sql.py:193), in _ida_query_NZPY(idadb, query, silent, first_row_only, autocommit)
    192 query = _prepare_query(query, silent)
--> 193 cursor.execute(query)
    195 if autocommit is True:

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849), in Cursor.execute(self, operation, args, stream)
    848        self._c.in_transaction = True
--> 849     self._c.execute(self, operation, args)
    850 except AttributeError as e:

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843), in Connection.execute(self, cursor, query, vals)
   1842 if self.error is not None:
-> 1843     raise ProgrammingError(self.error)    
   1845 return response

ProgrammingError: ERROR:  NZAUT-0005E: A model with the name '"ABCD"' cannot be found.


During handling of the above exception, another exception occurred:

ProgrammingError                          Traceback (most recent call last)
File [~/Documents/work/nzpyida/nzpyida/base.py:1873](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/base.py:1873), in IdaDataBase._drop(self, objectname, object_type)
   1872 try:
-> 1873     self._prepare_and_execute("DROP %s %s"%(to_drop,objectname))
   1874 except Exception as e:

File [~/Documents/work/nzpyida/nzpyida/base.py:2302](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/base.py:2302), in IdaDataBase._prepare_and_execute(self, query, autocommit, silent)
   2301 self._check_connection()
-> 2302 return sql._prepare_and_execute(self, query, autocommit, silent)

File [~/Documents/work/nzpyida/nzpyida/sql.py:60](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/sql.py:60), in _prepare_and_execute(idaobject, query, autocommit, silent)
     59 #print(query)
---> 60 cursor.execute(query)
     61 if autocommit is True:

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849), in Cursor.execute(self, operation, args, stream)
    848        self._c.in_transaction = True
--> 849     self._c.execute(self, operation, args)
    850 except AttributeError as e:

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843), in Connection.execute(self, cursor, query, vals)
   1842 if self.error is not None:
-> 1843     raise ProgrammingError(self.error)    
   1845 return response

ProgrammingError: ERROR:  relation does not exist TELCO2.ADMIN.DATA_FRAME_19039_1689241512


During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Cell In[21], line 2
      1 model = DecisionTreeClassifier(idadb, model_name='abcd')
----> 2 model.score(in_df=train_df, target_column='IS_FRAUD')

File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/classification.py:101](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/classification.py:101), in Classification.score(self, in_df, target_column, id_column)
     76 """
     77 Scores the model. The model must exist.
     78 
   (...)
     94     the model score
     95 """
     97 params = {
     98     'id': id_column
     99 }
--> 101 return self._score(in_df=in_df, predict_params=params, target_column=target_column)

File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:170](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:170), in PredictiveModeling._score(self, in_df, predict_params, target_column)
    168     return 1-res[0] if self.score_inv else res[0]
    169 finally:
--> 170     self.idadb.drop_table(out_table)
    171     if pred_view_needs_delete:
    172         self.idadb.drop_view(pred_view)

File [~/Documents/work/nzpyida/nzpyida/base.py:1208](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/base.py:1208), in IdaDataBase.drop_table(self, tablename)
   1179 def drop_table(self, tablename):
   1180     """
   1181     Drop a table in the database.
   1182 
   (...)
   1206         This operation cannot be undone if autocommit mode is activated.
   1207     """
-> 1208     return self._drop(tablename, "T")

File [~/Documents/work/nzpyida/nzpyida/base.py:1883](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/base.py:1883), in IdaDataBase._drop(self, objectname, object_type)
   1881 if self._is_netezza_system():
   1882     if "ERROR:  relation does not exist" in str(e.args[0]):
-> 1883         raise ValueError("Object does not exist.")
   1884     else:
   1885         raise e

ValueError: Object does not exist.
```

**Comment by mplpl** [comment]
This happens because as a cleanup we try to drop a temporary table that was never created. We should be smarter when doing it.

The right output in this case (that I got by quickly commenting that drop_table call [which is not a solution]) should look like this:

```
File [~/Documents/work/nzpyida/nzpyida/sql.py:85](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/sql.py:85), in ida_query(idadb, query, silent, first_row_only, autocommit)
     83     return _ida_query_ODBC_new(idadb, query, silent, first_row_only, autocommit)
     84 elif idadb._con_type == 'nzpy':
---> 85     return _ida_query_NZPY(idadb, query, silent, first_row_only, autocommit)
     86 else:
     87     return _ida_query_JDBC(idadb, query, silent, first_row_only, autocommit)

File [~/Documents/work/nzpyida/nzpyida/sql.py:193](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/sql.py:193), in _ida_query_NZPY(idadb, query, silent, first_row_only, autocommit)
    191 try:
    192     query = _prepare_query(query, silent)
--> 193     cursor.execute(query)
    195     if autocommit is True:
    196         idadb._autocommit()

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:849), in Cursor.execute(self, operation, args, stream)
    847        self._c.execute(self, "begin", None)
    848        self._c.in_transaction = True
--> 849     self._c.execute(self, operation, args)
    850 except AttributeError as e:
    851     if self._c is None:

File [/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/lib/python3.11/site-packages/nzpy-1.15-py3.11.egg/nzpy/core.py:1843), in Connection.execute(self, cursor, query, vals)
   1840 response = self.connNextResultSet(cursor)
   1842 if self.error is not None:
-> 1843     raise ProgrammingError(self.error)    
   1845 return response

ProgrammingError: ERROR:  NZAUT-0005E: A model with the name '"ABCD"' cannot be found.
```

**Comment by mplpl** [comment]
Fixed in 0.9.4

```
KeyError                                  Traceback (most recent call last)
Cell In[10], line 6
      4 x = DecisionTreeRegressor(idadb, model_name='SIMBOX_PREDICTxxx')
      5 with AutoDeleteContext(idadb):
----> 6     x.predict(in_df=df)

File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/regression_trees.py:199](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/regression_trees.py:199), in DecisionTreeRegressor.predict(self, in_df, out_table, id_column, variance)
    169 """
    170 Makes predictions based on this model. The model must exist.
    171 
   (...)
    191     the data frame containing row identifiers and predicted target values
    192 """
    194 params = {
    195     'id': q(id_column),
    196     'var': variance
    197     }
--> 199 return self._predict(in_df=in_df, params=params, out_table=out_table)

File [~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:111](https://file+.vscode-resource.vscode-cdn.net/Users/mpl/Documents/work/~/Documents/work/nzpyida/nzpyida/analytics/predictive/predictive_modeling.py:111), in PredictiveModeling._predict(self, in_df, params, out_table)
    108     raise TypeError("Argument in_df should be an IdaDataFrame")
    110 if not ModelManager(self.idadb).model_exists(self.model_name):
--> 111         raise KeyError("Model name not found in Model Manager, "
    112                     "use 'fit' function to train the model first")
    114 params['model'] = self.model_name
    115 return call_proc_df_in_out(proc=self.predict_proc, in_df=in_df, params=params,
    116     out_table=out_table)[0]

KeyError: "Model name not found in Model Manager, use 'fit' function to train the model first"
```
